File: impl/sdk/rs/resonate/src/effects.rs
use dashmap::DashMap;
use serde::Serialize;

use crate::codec::Codec;
use crate::error::Result;
use crate::send::Sender;
use crate::types::{PromiseCreateReq, PromiseRecord, PromiseSettleReq, PromiseState, SettleState};

/// The two durable operations the SDK needs. Built from Sender + Codec.
/// Maintains an internal cache of decoded PromiseRecords.
/// Shared via Arc — all contexts in an execution tree use the same instance.
pub struct Effects {
    sender: Sender,
    codec: Codec,
    cache: DashMap<String, PromiseRecord>,
}

impl Effects {
    /// Build Effects from a Sender, Codec, and optional preloaded promises.
    pub fn new(sender: Sender, codec: Codec, preload: Vec<PromiseRecord>) -> Self {
        let map = DashMap::new();
        for p in preload {
            if let Ok(decoded) = codec.decode_promise(p) {
                map.insert(decoded.id.clone(), decoded);
            }
        }

        Self {
            sender,
            codec,
            cache: map,
        }
    }

    /// Create a durable promise. Returns the decoded PromiseRecord.
    /// This is idempotent — if the promise already exists, it returns the existing record.
    pub async fn create_promise(&self, req: PromiseCreateReq) -> Result<PromiseRecord> {
        // 1. Check cache
        if let Some(cached) = self.cache.get(&req.id) {
            return Ok(cached.clone());
        }

        // 2. Encode param via codec
        let encoded_param = self.codec.encode(req.param.data_as_ref())?;
        let encoded_req = PromiseCreateReq {
            id: req.id.clone(),
            timeout_at: req.timeout_at,
            param: encoded_param,
            tags: req.tags.clone(),
        };

        // validation tracing
        let invocation = match encoded_req.tags.get("resonate:scope").map(String::as_str) {
            Some("local") => "run",
            Some("global") => "rpc",
            _ => "unknown",
        };
        tracing::info!(
            target: "resonate::validation",
            promise_id = %encoded_req.id,
            invocation,
            "promise_create_request"
        );

        // 3. Send request
        let record = self.sender.promise_create(encoded_req).await?;

        // 4. Decode response, cache, return
        let decoded = self.codec.decode_promise(record)?;
        self.cache.insert(decoded.id.clone(), decoded.clone());

        tracing::info!(
            target: "resonate::validation",
            promise_id = %decoded.id,
            invocation,
            state = ?decoded.state,
            "promise_create_response"
        );
        Ok(decoded)
    }

    /// Settle a durable promise with a result.
    pub async fn settle_promise<T: Serialize>(
        &self,
        id: &str,
        result: &Result<T>,
    ) -> Result<PromiseRecord> {
        // 1. Check cache — if already settled, return cached
        if let Some(cached) = self.cache.get(id) {
            if cached.state != PromiseState::Pending {
                return Ok(cached.clone());
            }
        }

        // 2. Build settle request
        let (state, value_data) = match result {
            Ok(val) => (SettleState::Resolved, serde_json::to_value(val)?),
            Err(err) => (SettleState::Rejected, crate::codec::encode_error(err)),
        };

        // 3. Encode value via codec
        let encoded_value = self.codec.encode(&value_data)?;

        let req = PromiseSettleReq {
            id: id.to_string(),
            state,
            value: encoded_value,
        };

        // 4. Send request
        tracing::info!(
            target: "resonate::validation",
            promise_id = %req.id,
            state = ?req.state,
            "promise_settle_request"
        );
        let record = self.sender.promise_settle(req).await?;

        // 5. Decode response, cache, return
        let decoded = self.codec.decode_promise(record)?;
        tracing::info!(
            target: "resonate::validation",
            promise_id = %decoded.id,
            state = ?decoded.state,
            "promise_settle_response"
        );
        self.cache.insert(decoded.id.clone(), decoded.clone());
        Ok(decoded)
    }
}

#[cfg(test)]
mod tests {
    use crate::test_utils::*;
    use crate::types::{PromiseCreateReq, PromiseState, Value};
    use std::collections::HashMap;

    // ── create_promise ──────────────────────────────────────────────

    #[tokio::test]
    async fn create_returns_cached_promise_from_preload_without_hitting_network() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![pending_promise("p1")]);

        let req = PromiseCreateReq {
            id: "p1".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        let record = effects.create_promise(req).await.unwrap();
        assert_eq!(record.state, PromiseState::Pending);
        assert_eq!(harness.get_send_count(), 0);
    }

    #[tokio::test]
    async fn create_hits_network_when_promise_not_in_preload() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![]);

        let req = PromiseCreateReq {
            id: "p2".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        let record = effects.create_promise(req).await.unwrap();
        assert_eq!(record.state, PromiseState::Pending);
        assert_eq!(harness.get_send_count(), 1);
    }

    #[tokio::test]
    async fn create_adds_to_cache_second_call_uses_cache() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![]);

        let req = PromiseCreateReq {
            id: "p3".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        effects.create_promise(req.clone()).await.unwrap();
        assert_eq!(harness.get_send_count(), 1);

        let record = effects.create_promise(req).await.unwrap();
        assert_eq!(record.state, PromiseState::Pending);
        assert_eq!(harness.get_send_count(), 1);
    }

    // ── settle_promise ──────────────────────────────────────────────

    #[tokio::test]
    async fn settle_returns_cached_when_already_settled_in_preload() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![resolved_promise("s1", serde_json::json!(42))]);

        let record = effects
            .settle_promise("s1", &Ok(serde_json::json!(99)))
            .await
            .unwrap();
        assert_eq!(record.state, PromiseState::Resolved);
        assert_eq!(harness.get_send_count(), 0);
    }

    #[tokio::test]
    async fn settle_hits_network_when_preloaded_promise_is_pending() {
        let harness = TestHarness::new();
        // Seed the stub so settle can find it
        harness.add_promise(pending_promise("s2")).await;
        let effects = harness.build_effects(vec![pending_promise("s2")]);

        let record = effects
            .settle_promise("s2", &Ok(serde_json::json!("ok")))
            .await
            .unwrap();
        assert_eq!(record.state, PromiseState::Resolved);
        assert_eq!(harness.get_send_count(), 1);
    }

    #[tokio::test]
    async fn settle_updates_cache_second_settle_is_cached() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![]);

        // Create the promise first
        let req = PromiseCreateReq {
            id: "s3".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        effects.create_promise(req).await.unwrap();
        assert_eq!(harness.get_send_count(), 1);

        // Settle it
        effects
            .settle_promise("s3", &Ok(serde_json::json!("done")))
            .await
            .unwrap();
        assert_eq!(harness.get_send_count(), 2);

        // Second settle should use cache
        let record = effects
            .settle_promise("s3", &Ok(serde_json::json!("done")))
            .await
            .unwrap();
        assert_eq!(record.state, PromiseState::Resolved);
        assert_eq!(harness.get_send_count(), 2);
    }

    #[tokio::test]
    async fn settle_hits_network_when_promise_not_in_cache() {
        let harness = TestHarness::new();
        // Seed stub directly (not in preload/cache)
        harness.add_promise(pending_promise("s4")).await;
        let effects = harness.build_effects(vec![]);

        let record = effects
            .settle_promise("s4", &Ok(serde_json::json!("ok")))
            .await
            .unwrap();
        assert_eq!(record.state, PromiseState::Resolved);
        assert_eq!(harness.get_send_count(), 1);
    }

    // ── cached promise values ───────────────────────────────────────

    #[tokio::test]
    async fn preloaded_pending_promise_has_decoded_param() {
        let harness = TestHarness::new();
        let param = serde_json::json!({"func": "f", "args": []});
        let effects = harness.build_effects(vec![pending_promise_with_param("v1", param.clone())]);

        let req = PromiseCreateReq {
            id: "v1".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        let record = effects.create_promise(req).await.unwrap();
        assert_eq!(record.param.data_or_null(), param);
    }

    #[tokio::test]
    async fn preloaded_resolved_promise_has_decoded_value() {
        let harness = TestHarness::new();
        let val = serde_json::json!({"answer": 42});
        let effects = harness.build_effects(vec![resolved_promise("v2", val.clone())]);

        let record = effects
            .settle_promise("v2", &Ok(serde_json::json!(0)))
            .await
            .unwrap();
        assert_eq!(record.value.data_or_null(), val);
    }

    #[tokio::test]
    async fn promise_created_via_network_has_correct_decoded_values() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![]);

        let param_data = serde_json::json!({"func": "myFunc", "args": [1, "two"]});
        let req = PromiseCreateReq {
            id: "v3".to_string(),
            timeout_at: i64::MAX,
            param: crate::types::Value {
                headers: None,
                data: Some(param_data.clone()),
            },
            tags: HashMap::new(),
        };
        effects.create_promise(req.clone()).await.unwrap();

        // Second call from cache
        let record = effects.create_promise(req).await.unwrap();
        assert_eq!(record.param.data_or_null(), param_data);
    }

    #[tokio::test]
    async fn promise_settled_via_network_has_correct_decoded_values() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![]);

        let req = PromiseCreateReq {
            id: "v4".to_string(),
            timeout_at: i64::MAX,
            param: Value::default(),
            tags: HashMap::new(),
        };
        effects.create_promise(req).await.unwrap();

        let val = serde_json::json!({"result": "success", "count": 7});
        effects
            .settle_promise("v4", &Ok(val.clone()))
            .await
            .unwrap();

        // Second settle from cache
        let record = effects
            .settle_promise("v4", &Ok(val.clone()))
            .await
            .unwrap();
        assert_eq!(record.value.data_or_null(), val);
    }

    #[tokio::test]
    async fn multiple_preloaded_promises_each_have_correct_values() {
        let harness = TestHarness::new();
        let effects = harness.build_effects(vec![
            pending_promise_with_param("m1", serde_json::json!({"func": "f", "args": []})),
            resolved_promise("m2", serde_json::json!("hello")),
            resolved_promise("m3", serde_json::json!([1, 2, 3])),
        ]);

        let r1 = effects
            .create_promise(PromiseCreateReq {
                id: "m1".to_string(),
                timeout_at: i64::MAX,
                param: Value::default(),
                tags: HashMap::new(),
            })
            .await
            .unwrap();
        assert_eq!(r1.state, PromiseState::Pending);
        assert_eq!(
            r1.param.data_or_null(),
            serde_json::json!({"func": "f", "args": []})
        );

        let r2 = effects
            .settle_promise("m2", &Ok(serde_json::json!(0)))
            .await
            .unwrap();
        assert_eq!(r2.state, PromiseState::Resolved);
        assert_eq!(r2.value.data_or_null(), serde_json::json!("hello"));

        let r3 = effects
            .settle_promise("m3", &Ok(serde_json::json!(0)))
            .await
            .unwrap();
        assert_eq!(r3.state, PromiseState::Resolved);
        assert_eq!(r3.value.data_or_null(), serde_json::json!([1, 2, 3]));

        assert_eq!(harness.get_send_count(), 0);
    }
}

File: impl/sdk/py/src/resonate/effects.py
from __future__ import annotations

from typing import TYPE_CHECKING, Literal, Protocol

from resonate.error import PlatformError, ResonateError, StoppedError
from resonate.observability import (
    Dropped,
    PromiseCreateRequested,
    PromiseCreateReturned,
    PromiseSettleRequested,
    PromiseSettleReturned,
    logging_observer,
)
from resonate.types import PromiseCreateReq, PromiseSettleReq

if TYPE_CHECKING:
    from resonate.codec import Codec
    from resonate.observability import Observer
    from resonate.send import PromiseFencing
    from resonate.types import PromiseRecord


class Effects(Protocol):
    """The two durable operations a :class:`~resonate.context.Context` performs.

    Deliberately *only* the two operations: the promise cache is an
    implementation concern of :class:`ResonateEffects`, not part of the
    contract, so a stand-in has two methods to write and nothing else.

    Implementations are expected to be idempotent (a cached or already-settled
    record short-circuits the network) and to behave as a circuit breaker: once
    a durable op has failed in this attempt, every later one raises rather than
    doing more work, so the task is released and re-delivery retries the whole
    execution.
    """

    async def create_promise(self, req: PromiseCreateReq) -> PromiseRecord: ...
    async def settle_promise[T](
        self, id: str, result: T | Exception
    ) -> PromiseRecord: ...


class ResonateEffects:
    """The two durable operations the SDK needs, built from a Sender and Codec.

    Maintains an internal cache of decoded :class:`PromiseRecord`s. A plain
    ``dict`` is safe here because the SDK runs single-threaded on asyncio --
    individual ``dict`` reads and writes are atomic, and no operation holds
    the cache across an ``await``.
    """

    def __init__(
        self,
        sender: PromiseFencing,
        codec: Codec,
        task_id: str,
        task_version: int,
        preload: list[PromiseRecord],
        observer: Observer = logging_observer,
    ) -> None:
        """Build Effects from a Sender, Codec, task lease, and preloaded promises.

        ``task_id``/``task_version`` are the active task's lease, used as the
        fencing token on every durable promise mutation. Each preloaded record
        is decoded into the cache; a record that fails to decode is skipped and
        reported to ``observer``.
        """
        self.sender = sender
        self.codec = codec
        self.task_id = task_id
        self.task_version = task_version
        self.cache: dict[str, PromiseRecord] = {}
        self._observer = observer
        self._stopped: bool = False
        for p in preload:
            self._absorb(p)

    def _absorb(self, record: PromiseRecord) -> None:
        """Decode and cache a server promise record.

        A record that fails to decode is skipped -- and reported as a
        :class:`~resonate.observability.Dropped` event, so the skip is a
        contract a test can assert rather than a silent ``return``. The decoded
        record is inserted monotonically (see ``_insert_monotonic``).
        """
        try:
            decoded = self.codec.decode_promise(record)
        except ResonateError as exc:
            self._observer(Dropped(what="preload-record", id=record.id, cause=str(exc)))
            return
        self._insert_monotonic(decoded)

    def _insert_monotonic(self, record: PromiseRecord) -> None:
        """Insert a decoded record, preserving monotonicity.

        Promise state is monotonic (pending -> terminal, then immutable), so a
        terminal cache entry is never overwritten by a (possibly stale) record.
        """
        existing = self.cache.get(record.id)
        if existing is not None and existing.state != "pending":
            return
        self.cache[record.id] = record

    async def create_promise(self, req: PromiseCreateReq) -> PromiseRecord:
        """Create a durable promise, returning the decoded record.

        Idempotent: a cached record (from preload or a prior call) is returned
        without touching the network.
        """
        if self._stopped:
            raise PlatformError([StoppedError()])

        cached = self.cache.get(req.id)
        if cached is not None:
            return cached

        try:
            encoded_req = PromiseCreateReq(
                id=req.id,
                timeout_at=req.timeout_at,
                param=self.codec.encode(req.param.data),
                tags=req.tags,
            )

            invocation = _invocation_of(encoded_req.tags.get("resonate:scope"))
            self._observer(
                PromiseCreateRequested(id=encoded_req.id, invocation=invocation)
            )

            res = await self.sender.task_fence_create(
                self.task_id, self.task_version, encoded_req
            )
            for p in res.preload:
                self._absorb(p)
            decoded = self.codec.decode_promise(res.promise)
        except ResonateError as exc:
            self._stopped = True
            raise PlatformError([exc]) from exc
        self._insert_monotonic(decoded)
        self._observer(
            PromiseCreateReturned(
                id=decoded.id, invocation=invocation, state=decoded.state
            )
        )
        return decoded

    async def settle_promise[T](self, id: str, result: T | Exception) -> PromiseRecord:
        """Settle a durable promise with a result.

        Idempotent: a cached non-pending record is returned without touching the
        network. ``result`` is the ``Result[T]`` to settle with -- a plain value
        resolves the promise, any ``Exception`` rejects it (the codec flattens it
        to the error shape, pickling the original where it can).
        """
        if self._stopped:
            raise PlatformError([StoppedError()])

        cached = self.cache.get(id)
        if cached is not None and cached.state != "pending":
            return cached

        state: Literal["resolved", "rejected"]
        state = "rejected" if isinstance(result, Exception) else "resolved"

        # Same conversion boundary as create_promise. Note this covers the
        # encode of the *user's* return value too: an unserializable return is
        # a PlatformError, so the task is released and re-delivered into the
        # same failure (a poison task) -- matching the pre-existing behavior of
        # the root-level encode in execute_until_blocked_inner.
        try:
            req = PromiseSettleReq(id=id, state=state, value=self.codec.encode(result))

            self._observer(PromiseSettleRequested(id=req.id, state=req.state))
            res = await self.sender.task_fence_settle(
                self.task_id, self.task_version, req
            )
            for p in res.preload:
                self._absorb(p)
            decoded = self.codec.decode_promise(res.promise)
        except ResonateError as exc:
            self._stopped = True
            raise PlatformError([exc]) from exc
        self._observer(PromiseSettleReturned(id=decoded.id, state=decoded.state))
        self._insert_monotonic(decoded)
        return decoded


def _invocation_of(scope: str | None) -> Literal["run", "rpc", "unknown"]:
    """Map a promise's ``resonate:scope`` tag to the invocation form that made it."""
    match scope:
        case "local":
            return "run"
        case "global":
            return "rpc"
        case _:
            return "unknown"
